plugin/pkg/scheduler/core/scheduling_queue.go

```diff
@@ -15,47 +15,679 @@ limitations under the License.
+	return f.FIFO.Delete(pod)
+}
+
 // Pop removes the head of FIFO and returns it.
 // This is just a copy/paste of cache.Pop(queue Queue) from fifo.go that scheduler
 // has always been using. There is a comment in that file saying that this method
 // shouldn't be used in production code, but scheduler has always been using it.
 // This function does minimal error checking.
-func (f *FIFO) Pop() (interface{}, error) {
+func (f *FIFO) Pop() (*v1.Pod, error) {
 	var result interface{}
 	f.FIFO.Pop(func(obj interface{}) error {
 		result = obj
 		return nil
 	})
-	return result, nil
+	return result.(*v1.Pod), nil
 }
 
-var _ = SchedulingQueue(&FIFO{}) // Making sure that FIFO implements SchedulingQueue.
+// FIFO does not need to react to events, as all pods are always in the active
+// scheduling queue anyway.
+func (f *FIFO) AssignedPodAdded(pod *v1.Pod)   {}
+func (f *FIFO) AssignedPodUpdated(pod *v1.Pod) {}
+
+// MoveAllToActiveQueue does nothing in FIFO as all pods are always in the active queue.
+func (f *FIFO) MoveAllToActiveQueue() {}
+
+// WaitingPodsOnNode returns pods that are nominated to run on the given node,
+// but FIFO does not support it.
+func (f *FIFO) WaitingPodsOnNode(nodeName string) []*v1.Pod {
+	return nil
+}
+
+func NewFIFO() *FIFO {
+	return &FIFO{FIFO: cache.NewFIFO(cache.MetaNamespaceKeyFunc)}
+}
+
+// UnschedulablePods is an interface for a queue that is used to keep unschedulable
+// pods. These pods are not actively reevaluated for scheduling. They are moved
+// to the active scheduling queue on certain events, such as termination of a pod
+// in the cluster, addition of nodes, etc.
+type UnschedulablePods interface {
+	Add(pod *v1.Pod)
+	Delete(pod *v1.Pod)
+	Update(pod *v1.Pod)
+	GetPodsForNode(nodeName string) []*v1.Pod
+	Get(pod *v1.Pod) *v1.Pod
+	Clear()
+}
+
+// PriorityQueue implements a scheduling queue. It is an alternative to FIFO.
+// The head of PriorityQueue is the highest priority pending pod. This structure
+// has two sub queues. One sub-queue holds pods that are being considered for
+// scheduling. This is called activeQ and is a Heap. Another queue holds
+// pods that are already tried and are determined to be unschedulable. The latter
+// is called unschedulableQ.
+// Heap is already thread safe, but we need to acquire another lock here to ensure
+// atomicity of operations on the two data structures..
+type PriorityQueue struct {
+	lock sync.RWMutex
+	cond sync.Cond
+
+	// activeQ is heap structure that scheduler actively looks at to find pods to
+	// schedule. Head of heap is the highest priority pod.
+	activeQ *Heap
+	// unschedulableQ holds pods that have been tried and determined unschedulable.
+	unschedulableQ *UnschedulablePodsMap
+	// receivedMoveRequest is set to true whenever we receive a move request, and
+	// is set to false, when we pop a pod. It indicates if we received a move
+	// request when a pod was in flight. In such case, we put the pod back into the
```

Comment:
The answer to the first question is yes. When scheduler tries to schedule pod A (when pod A is in flight), if we receive a move request to move pods to activeQ, we should probably put pod A to the activeQ if it is determined unschedulable. That's why this flag exists. The flag tracks any moves from unschedulableQ to the activeQ while scheduler attempts to schedule a pod. That's why popping a pod resets the flag to false. It marks the beginning of another cycle.
That said, you are right that the flag is not precise and there is no guarantee that we always see the next Pop() after pod A is put back into the queue.

Reply:
OK everything made sense except the last sentence. What do you mean by "we always see"?

Reply:
Scheduler updates the status of an unschedulable pod at the end of each cycle, but since receiving the event for the update happens asynchronously (via event handlers), scheduler may Pop() another pod and then the event for adding the pod which was picked in the previous cycle arrives. That's what I meant by the last sentence.